### DISCUSSION TITLE
What would you like to see in a project template?

### DISCUSSION THREAD

**Discussion by CleanCut** [bevyengine/bevy#557] [Q&A]
#### Long version:

I've created the [`bevy_template`](https://github.com/CleanCut/bevy_template) project to make it easier to create a Bevy project and not have to configure it from scratch.

Right now it's mostly focused on configuring things to enable fast compilation, and it works best (has the most features) on macOS -- because I use macOS.

I'm asking partly because I want to try out this beta "Discussions" feature, and partly because I really want to know what folks would find useful in a project template. 😁

**Comment by mockersf** [comment]
Not about the content of the template, but why not use a project generator ?
At work we are using [ffizer](https://github.com/ffizer/ffizer) but there are a [few others](https://github.com/ffizer/ffizer/wiki/Alternatives).

**Reply by davidB** [reply]
I'm the author of [ffizer](https://github.com/ffizer/ffizer), and I'll be happy to help (includes add missing feature, explain, add missing doc or sample to ffizer). 2 interesting features of ffizer:
- template can be composed by other template
- template can be used to update an existing folder (show diff, basic merge)

**Reply by CleanCut** [reply]
This is very intriguing!  I may look into this. 😄

**Comment by will-hart** [comment]
I used yours to get up and running (more of a copy paste of the structure as I'm on windows) but for the most part I'm building up my own base project which I can use as a quick start without rewriting a bunch of common things. 

It currently has workspace crates for: 

- transitioning between different "scenes" (not using the scenes feature) 
 - tracking loading progress (ie can do a loading screen while assets load)
- an slightly different RNG like an RPG style quasi random generator
- helpers for animated sprites
- a very basic "buffable" character stats system
- other random poorly written code

I'm currently looking at a way to write a generic data loader for RON data files to load game data, but I'm not sure it will be as easy to support using the existing asset loader pipeline as is hoped. I might submit a PR to bevy for this.

**Reply by CleanCut** [reply]
That's an interesting idea! We could offer some sort of menu of curated plugins to include and maybe even wire them up for the user.

**Comment by zicklag** [comment]
It seems like it could be useful to have specific templates for different kinds of games. Like a "deploy this template and you'll have a working top-down adventure-style character walking around a small map". Then a person can deploy the template and start *modifying* to learn how things work, which sometimes is a lot easier to do that to start *making* and can be done without a ton of guidance depending on how simple the example is or how easy the engine is to understand at all.

We cold have templates for stuff like:

  - top down adventure game
  - top down shoot-em-up space shooter or something
  - side scrolling platformer

They could be super minimal, but just enough to give the user an idea of how to use it for a certain purpose, or at least to get started.

---

Not sure how much of a demand there is for this, just a thought.

**Reply by CleanCut** [reply]
I like that idea! This dovetails nicely into the "use a project generator" suggestion.

I'm (slowly) working with a collaborator to make a bunch of demo projects.  We're not planning on making them fancy enough that I would consider them "showcases", but they could certainly be considered starter templates.  Perhaps that's a good way to tie them together - make it so the template tool/generator/whatever could let you choose one if you like.

**Reply by denormative** [reply]
It would be good if they could be tied into the CI/testing side of things as well. The biggest problem I had with playing with unity a few years ago, and godot more recently, was finding non-trivial examples that worked without spending hours just trying to debug them.

So something like the trivial breakout example included with the code base would work, but once you want to try to understand a 'real game' (like, say, the breakout example with a main menu and scoring with win/loss conditions), you'd be left poking around github repos or even store bought assets where it wouldn't work with the current version and the operating systems you're using, or other annoyances.

Anyway, given how fast moving bevy is at the moment it seems like a good idea anyway, just my two cents. :)
